primitives/src/circuit/merkle_tree/mod.rs

```diff
@@ -75,10 +77,133 @@ where
 
     /// Enforce correct `merkle_proof` for the `elem` against
     /// `expected_merkle_root`.
-    fn enforce_merkle_proof(
+    fn enforce_membership_proof(
         &mut self,
         elem: Self::LeafVar,
         merkle_proof: Self::MerklePathVar,
         expected_merkle_root: Variable,
     ) -> Result<(), CircuitError>;
 }
+
+/// Gadget for the sparse merkle tree
+pub trait SparseMerkleTreeGadget<M>: MerkleTreeGadget<M>
```

Comment:
Is it better to call it `UniversalMerkleTreeGadget`?
There was some discussion before, where we concluded that non_membership_proof is a feature of Universal Merkle tree.

Reply:
Ah sure, I think I either missed the discussion or it happened before my time :)

Reply:
Done: https://github.com/EspressoSystems/jellyfish/pull/172/commits/c6fae19db16356b33ac0c8ab09ecc72fb61adc5a